File: modules/audio-eq/android/src/main/java/expo/modules/audioeq/AudioEQModule.kt
package expo.modules.audioeq

import android.media.audiofx.Equalizer
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class AudioEQModule : Module() {
  private var equalizer: Equalizer? = null
  private var bandGains = mutableMapOf<Int, Short>()
  private var isEnabled = true

  override fun definition() = ModuleDefinition {
    Name("AudioEQ")

    AsyncFunction("initialize") {
      setupEqualizer()
    }

    AsyncFunction("setBandGain") { band: Int, gain: Double ->
      setGain(band, gain.toFloat())
    }

    AsyncFunction("getBandCount") {
      equalizer?.numberOfBands?.toInt() ?: 5
    }

    AsyncFunction("getBandInfo") { band: Int ->
      val eq = equalizer ?: return@AsyncFunction mapOf<String, Any>()
      if (band < 0 || band >= eq.numberOfBands) return@AsyncFunction mapOf<String, Any>()

      val freqRange = eq.bandFreqRange(band.toShort())
      val centerFreq = (freqRange[0] + freqRange[1]) / 2
      val minGain = eq.bandLevelRange[0].toDouble() / 100.0
      val maxGain = eq.bandLevelRange[1].toDouble() / 100.0

      mapOf(
        "index" to band,
        "centerFreq" to (centerFreq.toDouble() / 1000.0),
        "minGain" to minGain,
        "maxGain" to maxGain,
      )
    }

    AsyncFunction("setEnabled") { enabled: Boolean ->
      isEnabled = enabled
      equalizer?.enabled = enabled
    }

    AsyncFunction("teardown") {
      teardownEqualizer()
    }

    OnDestroy {
      teardownEqualizer()
    }
  }

  private fun setupEqualizer() {
    teardownEqualizer()
    try {
      val eq = Equalizer(0, 0)
      eq.enabled = isEnabled
      equalizer = eq
    } catch (e: Exception) {
      // Audio session not available yet
    }
  }

  private fun setGain(band: Int, gain: Float) {
    val clampedGain = (-12.0f).coerceAtLeast(gain.coerceAtMost(12.0f))
    bandGains[band] = (clampedGain * 100).toInt().toShort()
    equalizer?.setBandLevel(band.toShort(), bandGains[band] ?: 0)
  }

  private fun teardownEqualizer() {
    equalizer?.release()
    equalizer = null
  }
}

File: modules/audio-eq/ios/AudioEQModule.swift
import ExpoModulesCore
import AVFoundation

public class AudioEQModule: Module {
  private var audioEngine: AVAudioEngine?
  private var eqNode: AVAudioUnitEQ?
  private var bandGains: [Int: Float] = [:]
  private var isEnabled: Bool = true

  private static let bandFrequencies: [(index: Int, centerFreq: Float)] = [
    (0, 60),
    (1, 230),
    (2, 910),
    (3, 3600),
    (4, 14000),
  ]

  public func definition() -> ModuleDefinition {
    Name("AudioEQ")

    AsyncFunction("initialize") { () -> Void in
      self.setupEngine()
    }

    AsyncFunction("setBandGain") { (band: Int, gain: Double) -> Void in
      self.setGain(band: band, gain: Float(gain))
    }

    AsyncFunction("getBandCount") { () -> Int in
      return AudioEQModule.bandFrequencies.count
    }

    AsyncFunction("getBandInfo") { (band: Int) -> [String: Any] in
      guard band >= 0 && band < AudioEQModule.bandFrequencies.count else {
        return [:]
      }
      let info = AudioEQModule.bandFrequencies[band]
      return [
        "index": info.index,
        "centerFreq": info.centerFreq,
        "minGain": -12.0,
        "maxGain": 12.0,
      ]
    }

    AsyncFunction("setEnabled") { (enabled: Bool) -> Void in
      self.isEnabled = enabled
      self.eqNode?.bypass = !enabled
    }

    AsyncFunction("teardown") { () -> Void in
      self.teardownEngine()
    }

    OnDestroy {
      self.teardownEngine()
    }
  }

  private func setupEngine() {
    teardownEngine()

    let engine = AVAudioEngine()
    let eq = AVAudioUnitEQ(numberOfBands: AudioEQModule.bandFrequencies.count)

    for bandInfo in AudioEQModule.bandFrequencies {
      let band = eq.bands[bandInfo.index]
      band.filterType = .parametric
      band.frequency = bandInfo.centerFreq
      band.bandwidth = 1.0
      band.gain = bandGains[bandInfo.index] ?? 0.0
      band.bypass = false
    }

    engine.attach(eq)
    engine.connect(eq, to: engine.mainMixerNode, format: nil)

    do {
      try engine.start()
      self.audioEngine = engine
      self.eqNode = eq
    } catch {
      print("AudioEQ: Failed to start engine: \(error)")
    }
  }

  private func setGain(band: Int, gain: Float) {
    bandGains[band] = gain
    let clampedGain = max(-12.0, min(12.0, gain))
    guard let eq = eqNode, band >= 0 && band < eq.bands.count else { return }
    eq.bands[band].gain = clampedGain
  }

  private func teardownEngine() {
    audioEngine?.stop()
    if let eq = eqNode {
      audioEngine?.detach(eq)
    }
    audioEngine = nil
    eqNode = nil
  }
}
